### DISCUSSION TITLE
Linux/UNIX build for NuGet.CommandLine

### DISCUSSION THREAD

**Discussion by YakoYakoYokuYoku** [NuGet/Home#10603] [Ideas]
Long ago the `nuget` CLI could be built using Mono `xbuild`. But as time passed and `xbuild` wasn't making the cut and it was trimmed off the repository while leaving no documented way of building it for Linux or UNIX-like distros.

Time has passed since with no news about it until recently. Back in February of the past year I published a [patch in the Solus Phabricator](https://dev.getsol.us/D8310) that updated it to a much newer version of NuGet (latest version of the package was `2.11`, still supporting `xbuild`). Although I haven't updated it since then because of failing builds with `mono-msbuild`, I'll now explain how I've built version `5.10.0.7131`.

Basically the steps are:

- Restore both `build/bootstrap.proj` and `src/NuGet.Clients/NuGet.CommandLine.csproj`.
- Build with `dotnet` `src/NuGet.Clients/NuGet.CommandLine.csproj` with the `SkipILMergeOfNuGetExe` property set to `true`.
- Locate `NuGet.exe` and its `*.dll` dependencies and install it to the proper dir or...
- Use `dotnet publish -o <your output path>`.
- Remove `Microsoft.VisualStudio.Setup.Configuration.Interop.dll` to avoid licensing issues.

`mono` runs `NuGet.exe` as intended although I couldn't find the way to do it with `dotnet` yet.

As an alternative to `ilmerge.exe` (as it uses native Windows libs) I'd suggest to produce [self-contained applications](https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file).

**Comment by zivkan** [comment]
@YakoYakoYokuYoku Have you tried using the `dotnet` CLI?  If it's missing commands that `nuget.exe` has, which commands do you need? If the dotnet cli added those commands, would you still want nuget.exe?

Unfortunately self-contained .NET Core apps are not a direct replacement for ILMerged .NET Framework apps. .NET Core apps, during the publish step, need to select a runtime identifier (RID). This makes the publish output specific to Windows, Linux or Mac, and therefore a single executable is no longer compatible with all platforms like a .NET Framework assembly is thanks to mono.

Hence, I advocate to make NuGet's features in the dotnet CLI a superset of nuget.exe, and then we can make nuget.exe obsolete. There are still some gaps, but if customers give us feedback about which commands are most important to them, we can prioritize the most requested commands.

**Reply by YakoYakoYokuYoku** [reply]
> Have you tried using the dotnet CLI? If it's missing commands that nuget.exe has, which commands do you need? If the dotnet cli added those commands, would you still want nuget.exe?

In my case `dotnet nuget` was lacking `nuget`'s `install` subcommand, but I can open an issue in the proper `dotnet` repo. My guide was more directed at projects that still use `nuget.exe` like Mono or [Godot](https://github.com/godotengine/godot). Still I also believe that the `dotnet` CLI is the way to go.

**Reply by zivkan** [reply]
The NuGet team owns the experience, so if you create it in the dotnet/SDK repo, they'll move it back to NuGet/Home. I had a quick search, and I don't see an existing request for this command to suggest for you to upvote, so creating a new feature request would be the best option here.
